scipy/sparse/csgraph/_flow.pyx

```diff
@@ -670,3 +683,943 @@ cdef ITYPE_t[:] _dinic(
+    The cycle is oriented in the direction from p to q.
+    """
+    cdef:
+        ITYPE_t num_edges, num_edges_R, idx
+        ITYPE_t offset, Wn_len, We_len, tmp
+
+    w = _find_apex(p, q, subtree_size, parent)
+    num_edges = _trace_path(p, w, parent,
+                            parent_edge,
+                            Wn, We)
+    Wn_len = num_edges + 1
+    We_len = num_edges
+
+    num_edges_R = _trace_path(q, w, parent,
+                              parent_edge,
+                              WnR, WeR)
+    for idx in range(We_len//2):
+        We[idx], We[We_len - idx - 1] = We[We_len - idx - 1], We[idx]
+    for idx in range(Wn_len//2):
+        Wn[idx], Wn[Wn_len - idx - 1] = Wn[Wn_len - idx - 1], Wn[idx]
+    if ((We_len == 1 and We[0] != i) or We_len != 1):
+        We[We_len] = i
+        We_len += 1
+    for idx in range(num_edges_R):
+        Wn[idx + Wn_len] = WnR[idx]
+        We[idx + We_len] = WeR[idx]
+    Wn_len += num_edges_R
+    We_len += num_edges_R
+    Wne_len[0] = Wn_len
+    Wne_len[1] = We_len
+
+cdef inline ITYPE_t _residual_capacity(
+        ITYPE_t i,  # IN
+        ITYPE_t p,  # IN
+        ITYPE_t[:] edge_sources,  # IN
+        ITYPE_t[:] edge_capacities,  # IN
+        ITYPE_t[:] edge_flow  # IN
+        ) nogil:
+    """
+    Returns the residual capacity of an edge i in the direction away
+    from its endpoint p.
+    """
+    if edge_sources[i] == p:
+        return edge_capacities[i] - edge_flow[i]
+    else:
+        return edge_flow[i]
+
+cdef edge_result _find_leaving_edge(
+        ITYPE_t[:] Wn,  # IN
+        ITYPE_t[:] We,  # IN
+        ITYPE_t[:] Wne_len,  # IN
+        ITYPE_t[:] edge_sources,  # IN
+        ITYPE_t[:] edge_targets,  # IN
+        ITYPE_t[:] edge_capacities,  # IN
+        ITYPE_t[:] edge_flow,  # IN
+        edge_result ret_values  # IN/OUT
+        ) nogil:
+    """
+    Returns the leaving edge in a cycle represented by Wn and We.
+    """
+    cdef:
+        ITYPE_t min_res_cap, res_cap
+        ITYPE_t i = -2, p, j = -2, s = -2, t, idx_e, idx_n
+    min_res_cap = INT_MAX
+    idx_n = Wne_len[0] - 1
+    idx_e = Wne_len[1] - 1
+    while idx_n >= 0 and idx_e >= 0:
+        i = We[idx_e]
+        p = Wn[idx_n]
+        res_cap = _residual_capacity(i, p,
+                                     edge_sources,
+                                     edge_capacities,
+                                     edge_flow)
+        if res_cap < min_res_cap:
+            min_res_cap = res_cap
+            j = i
+            s = p
+        idx_n -= 1
+        idx_e -= 1
+    if edge_sources[j] == s:
+        t = edge_targets[j]
+    else:
+        t = edge_sources[j]
+    ret_values.i = j
+    ret_values.p = s
+    ret_values.q = t
+    return ret_values
+
+cdef inline ITYPE_t _index(
+        ITYPE_t[:] W,  # IN
+        ITYPE_t W_len,  # IN
+        ITYPE_t i  # IN
+        ) nogil:
+    """
+    Returns the index of value in the given array.
+    """
```

Comment:
Isn't a imperative form generally used? Also, couldn't this be a one-liner docstring?
```suggestion
    """Return the index of value in the given array."""
```

The same suggestion applies in other docstring.